core/src/main/java/org/geysermc/geyser/session/GeyserSession.java

```diff
@@ -1548,6 +1552,25 @@ public void sendCommand(String command) {
         this.sendCommandPacket(command);
     }
 
+    @Override
+    public @NonNull String clientConnectionAddress() {
+        if (clientData == null) {
+            return "";
```

Comment:
Not sure how heavily we're using it these days, but this almost feels like a perfect situation for an Optional.

Reply:
Or just make it nullable ^^